src/gas/fetchGasEstimatesViaEthFeeHistory.ts

```diff
@@ -0,0 +1,167 @@
+    return block.priorityFeesByPercentile[settings.percentile];
+  });
+  const medianPriorityFee = medianOf(priorityFees);
+  const adjustedPriorityFee =
+    medianPriorityFee * (1 - settings.priorityFeeReduction);
+  const suggestedMaxPriorityFeePerGas = Math.max(
+    adjustedPriorityFee,
+    settings.minSuggestedMaxPriorityFeePerGas,
+  );
+  const suggestedMaxFeePerGas = adjustedBaseFee + suggestedMaxPriorityFeePerGas;
+
+  return {
+    ...settings.estimatedWaitTimes,
+    suggestedMaxPriorityFeePerGas: weiToGwei(
+      suggestedMaxPriorityFeePerGas,
+    ).toString(),
+    suggestedMaxFeePerGas: weiToGwei(suggestedMaxFeePerGas).toString(),
+  };
+}
+
+/**
+ * Calculates a set of estimates suitable for different priority levels based on the data returned
+ * by `eth_feeHistory`.
+ *
+ * @param latestBaseFeePerGas - The base fee per gas recorded for the latest block.
+ * @param blocks - More information about blocks we can use to calculate estimates.
+ * @returns The estimates.
+ */
+function calculateGasEstimatesForAllPriorityLevels(
+  latestBaseFeePerGas: number,
+  blocks: { priorityFeesByPercentile: Record<Percentile, number> }[],
+) {
+  return PRIORITY_LEVELS.reduce((obj, priorityLevel) => {
+    const gasEstimatesForPriorityLevel = calculateGasEstimatesForPriorityLevel(
+      priorityLevel,
+      latestBaseFeePerGas,
+      blocks,
+    );
+    return { ...obj, [priorityLevel]: gasEstimatesForPriorityLevel };
+  }, {} as Pick<GasFeeEstimates, PriorityLevel>);
+}
+
+/**
+ * Generates gas fee estimates based on gas fees that have been used in the recent past so that
+ * those estimates can be displayed to users.
+ *
+ * To produce the estimates, the last 5 blocks are read from the network, and for each block, the
+ * priority fees for transactions at the 10th, 20th, and 30th percentiles are also read (here
+ * "percentile" signifies the level at which those transactions contribute to the overall gas used
+ * for the block, where higher percentiles correspond to higher fees). This information is used to
+ * calculate reasonable max priority and max fees for three different priority levels (higher
+ * priority = higher fee).
+ *
+ * @param ethQuery - An EthQuery instance.
+ * @returns Base and priority fee estimates, categorized by priority level, as well as an estimate
+ * for the next block's base fee.
+ */
+export default async function fetchGasEstimatesViaEthFeeHistory(
+  ethQuery: EthQuery,
+): Promise<GasFeeEstimates> {
+  const feeHistory = await fetchFeeHistory<Percentile>({
+    ethQuery,
+    numberOfBlocks: NUMBER_OF_RECENT_BLOCKS,
+    percentiles: PERCENTILES,
+  });
+  const latestBlock = feeHistory.blocks[feeHistory.blocks.length - 1];
+  const latestBaseFeePerGas = latestBlock.baseFeePerGas;
+  const levelSpecificGasEstimates = calculateGasEstimatesForAllPriorityLevels(
+    latestBaseFeePerGas,
+    feeHistory.blocks,
+  );
+  const estimatedBaseFee = weiToGwei(latestBlock.baseFeePerGas).toString();
```

Comment:
I converted this to a string to copy what we do in `weiHexToGweiDec` (and elsewhere), but is it really that important to do this? We already have a floating point number, so I'm not sure what the benefit here is.

Reply:
Update: According to the discussion [here](https://consensys.slack.com/archives/G1L7H42BT/p1635528922064800) we should probably just leave this as a BN.